Latest patient bill

lomamech opened the pull request:
The latest bill should show the most recent bill of a patient
    Showing the most recent bill
- Update inteface for patient record
- Adding routes for latest bill
- Update patient service with service latest
- Update controllers patients checkin.js
- Integration test for Latest patient Bill
  @jniles

lomamech committed:
Showing the most recent bill
Update inteface for patient record
Adding routes for latest bill
Update patient service with service latest
Update controllers patients checkin.js

lomamech committed:
Integration test for Latest patient Bill

jniles commented on client/src/i18n/fr.json:
```diff
@@ -1017,8 +1025,12 @@
         "TOTAL_ASC" : "Le plus grand"
       },
       "TEXT"               : "Text",
+      "THIS_INVOICE_1"     : "Cette facture n'a pas été compensée en ",
```
A better `i18n` key for this would be "UNPAID_INVOICE_1".

jniles commented on client/src/partials/patients/record/patient_record.js:
```diff
@@ -46,6 +46,13 @@ function PatientRecordController($stateParams, Patients, Notify, moment, Upload,
+      // get the latest Invoice
+      Patients.latest(vm.patient.debtor_uuid)
+        .then(function (result){
+          vm.patientInvoice = result[0];          
```
For the latest invoice, you will always want a single invoice (or none, if the patient was never billed).  Instead of sending an array, you should send a single JSON object to the client.  That way you will not have to do `result[0]`.

jniles commented on server/config/routes.js:
```diff
@@ -379,6 +379,7 @@ exports.configure = function configure(app) {
 
   app.get('/patients/:uuid/visits', patients.checkin.list);
   app.post('/patients/:uuid/checkin', patients.checkin.create);
+  app.get('/patients/:uuid/invoices/latest', patients.checkin.latest);
```
Hmmm...  `patients.checkin.latest` makes it seem like the latest patient checkin.  Could you put this code somewhere else?   For example, in `patientInvoices.latest` or `patients.latestInvoice`?

jniles commented on server/controllers/medical/patients/checkin.js:
```diff
@@ -102,3 +104,81 @@ function create(req, res, next) {
+      JOIN project ON invoice.project_id = project.id `;
+
+    var sql2 =
+      `SELECT COUNT(BUID(i.uuid)) as numberPayment
```
Instead of using `BUID()` in the SQL statement, use the javascript function `db.bid()`.  The `BUID()` function in SQL converts from BINARY to STRING, while the `db.bid()` converts from STRING to BINARY.

jniles commented on server/controllers/medical/patients/checkin.js:
```diff
@@ -102,3 +104,81 @@ function create(req, res, next) {
+  })
+  .then(function (results) {
+    var invoices = results[0];
+    var numberPayment = results[1][0].numberPayment;
```
This is pretty ugly code.  One way to clean this up would [use the `.spread()` operator on promises](https://github.com/kriskowal/q#combination).

Like this:

``` js
return q.all([execSql, execSql2]);
.spread(function (invoices, payments) {
  var numberPayment = payments[0].numberPayment;
})
```

jniles commented on test/integration/patientCheckIn.js:
```diff
@@ -37,4 +39,19 @@ describe('(/patients/:uuid/visits) Patient Check In', () => {
+      .then(function (result) {
+        expect(result).to.have.status(200);
+
+        expect(result.body[0]).to.have.keys('uid', 'reference', 'credit', 'debit', 
```
To make this test stronger, check that the `debtor_uuid` is the same as the debtorUuid requested.

jniles commented on client/src/partials/patients/record/patient_record.html:
```diff
@@ -117,13 +117,78 @@
           <div class="panel panel-default">
             <div class="panel-heading">
-              <span class="glyphicon glyphicon-file"></span> Latest Bill
+              <span class="glyphicon glyphicon-file"></span>
```
Could you make this either a component, or a separate controller, like the other two panels?

lomamech committed:
Using UNPAID_INVOICE instead THIS_INVOICE

lomamech committed:
Update and fixes of review
Changing the root for latest invoice
Update and improve the integration test for latest invoice
Using a component bhLatestInvoice for the client side

lomamech committed:
Improve test integration for Latest Patient invoice

lomamech committed:
Sanitaze code
Adding moment in $inject
Remove unused Variable

jniles commented on client/src/js/components/bhLatestInvoice.js:
```diff
@@ -0,0 +1,52 @@
+  }
+});
+
+LatestInvoice.$inject = [
```
You should `$inject` `moment` here, so that the lint tests pass.

jniles commented on client/src/js/components/bhLatestInvoice.js:
```diff
@@ -0,0 +1,52 @@
+  var vm = this;
+
+  /** global variables */
+  vm.session = {
```
It doesn't look like you need any of these variables...  what does `enableSearch` do?  Please do not just copy code - the code must mean something.

jniles commented on client/src/partials/patients/record/patient_record.html:
```diff
@@ -115,18 +115,17 @@
-            </div>
-            <div class="panel-body" style="height : 250px;">
+          <!-- Latest Invoice  -->
+          <bh-latest-invoice
```
This is the right idea.

jniles commented on client/src/partials/templates/bhLatestInvoice.tmpl.html:
```diff
@@ -0,0 +1,81 @@
+  <div class="panel-body" style="height : 250px;">
+
+    <div style="text-align : left; font-size : 1em; ">
+      <table ng-if="$ctrl.session.patientInvoice" style="width:100%; vertical-align:top">
```
You actually do not need `session`.  Since you are using `controllerAs`, there is no need to introduce a second object `$ctrl.session` ... you can just bind directly to `$ctrl`.

jniles commented on server/controllers/medical/patients/checkin.js:
```diff
@@ -21,6 +21,7 @@ const db    = require('../../../lib/db');
 const topic = require('../../../lib/topic');
 
 const BadRequest = require('../../../lib/errors/BadRequest');
+const q     = require('q');
```
Do you need `q` here?  I think it may be left over some other change.

jniles commented on server/config/routes.js:
```diff
@@ -379,6 +379,7 @@ exports.configure = function configure(app) {
 
   app.get('/patients/:uuid/visits', patients.checkin.list);
   app.post('/patients/:uuid/checkin', patients.checkin.create);
+  app.get('/patients/:uuid/invoices/latest', patients.latestInvoice);
```
Much better!

jniles commented on server/controllers/medical/patients/index.js:
```diff
@@ -553,3 +556,89 @@ function subsidies(req, res, next) {
+
+    return q.all([execSql, execSql2]);
+  })
+  .spread(function (invoices, payments){
```
Much cleaner!

jniles commented on test/integration/patientInvoice.js:
```diff
@@ -55,6 +57,18 @@ describe('The /invoices API', function () {
+        expect(result).to.have.status(200);
+        expect(result).to.be.json;
+        expect(result.body).to.have.keys('uid', 'reference', 'credit', 'debit', 'balance', 'entity_uuid', 'uuid', 'debtor_uuid', 'user', 'date', 'cost', 'numberPayment');        
+        expect(result.body.entity_uuid).to.equal(debtorUuid);
```
This is a much stronger test.

jniles commented on test/integration/patientInvoice.js:
```diff
@@ -55,6 +57,18 @@ describe('The /invoices API', function () {
+        expect(result.body).to.have.keys('uid', 'reference', 'credit', 'debit', 'balance', 'entity_uuid', 'uuid', 'debtor_uuid', 'user', 'date', 'cost', 'numberPayment');        
+        expect(result.body.entity_uuid).to.equal(debtorUuid);
+      })
+      .catch(helpers.api.help);
```
I don't think `helpers.api.help` exists...

lomamech committed:
Sanitaze
Remove no necessary property variable session
Adding property invoiceslength
Update test integration and sanitaze

lomamech committed:
Remove unused q in controllers patient checkin

jniles commented on bower.json:
```diff
@@ -42,6 +42,7 @@
   "resolutions": {
-    "angular": "1.5.8"
+    "angular": "1.5.8",
+    "Chart.js": "2.1.6"
```
Could you remove this resolution?  It will affect the build.

lomamech committed:
Fix error in sql clause using debtor_uuid instead invoice_uuid

lomamech committed:
Remove charte in resolutions in bower.json

lomamech committed:
Sanitaze code and using npm install
